docs/madr/decisions/046-grpc-tcp-probes.md

```diff
@@ -0,0 +1,289 @@
+
+### Proposed option: option 1 - Transform gRPC probes to HTTP probes, merge all the HTTP probes by one HTTP listener
+
+Transform user defined gRPC probes into HTTP probes first and then use the existing Virtual Probe HTTP listener to handle incoming probe traffic. This needs n new intermediate layer translating HTTP probes back into gRPC probes (the translator), a reasonable place is put it in the kuma-dp.
+
+Handling incoming probe request from kubelet:
+
+1. If the user defines an HTTP probe, requests from kubelet are forwarded/redirected to application directly with necessary path/port rewriting, etc.
+
+2. If the user defines a gRPC probe, we'll also receive HTTP requests from kubelet, since they are transformed to HTTP probes. We forward these requests to the translator and the translator is responsible for handling the probe request, by detecting the actual status from the application.
+
+The whole workflow will be like the following diagram:
+
+```
+  kubelet (TCP Probes)                                   HTTP to TCP translator  --->  Application (TCP port)
+               ↘                                                ↗ 
+                 Pod (HTTP Probes)  --->  Virtual Probe Listener (HTTP routing)
+               ↗                                                ↘
+  kubelet (gRPC Probes)                                  HTTP to gRPC translator  --->  Application (gRPC healthcheck)
+```
+
+#### Positive Consequences
+
+- Don't need to introduce new ports for gRPC probes
+
+- Better user experience: don't need to introduce new annotations, the user can still specify the port as they are doing now.
+
+- Less virtual listeners on sidecar and saves a port to be taken from application: the translator will be used to handle both TCP probes and gRPC probes
+
+- Don't need to introduce new fields on the `Dataplane` object
+
+#### Negative Consequences
+
+- Requires an extra component (the translator) in `kuma-dp`, which increases the implementation complexity
+
+- Less intuitiveness, harder for troubleshooting issues
+
+#### Implementation details
+
+The probes support can be implemented in the following steps:
+
+1. Change the `Dataplane` model to contain fields for gRPC and TCP probes (see examples below)
+2. Create a new component in `kuma-dp` (the probe translator) to handle the incoming HTTP probes by performing the actual gRPC and TCP probes and returning the result
+3. Generate a new Envoy cluster `kuma-virtual-probes` and point it to the component
+4. Make sidecar injector to:
+   1. Get the correct port for the probe translator, and generate corresponding command line arguments for `kuma-dp` 
+   2. Identify gRPC and TCP probes and transform them into the updated `Dataplane` object
+5. Change the route of Virtual Probe listener (in `probe_generator`) and forward the user defined gRPC and TCP probes into the new component in `kuma-dp`
```

Comment:
What I'm reading here is that the path is
kubelet -> Envoy -> kuma-dp -> app.

Why involve Envoy if the logic is in kuma-dp? We could just do kubelet -> kuma-dp -> app.

This way we get rid of any config in Dataplane and get rid of xds generation

Reply:
If we don't involve Envoy, we'll need to skip the port when setting up the traffic redirection (using IPtables, etc.)

Reply:
which does not sound like a big deal maybe?
Sounds like fewer broken changes than allocating a new port.

Reply:
Yeah, agreed when we think in this way.
@bartsmykla WDYT?

Reply:
I'll mark this conversation as resolved since it's now in the proposed approach.